Continued work on updating to Ember-Simple-Auth 1.0

acorncom opened the pull request:
This continues the work on #89 slightly

Commit:
Updated to ember-simple-auth 1.0

wayne-o commented:
Just pulled this and ran ember test and all tests passed?

wayne-o commented:
OK npm test shows the failures - having a look now - not sure I totally understand what's going on though - will do my best

acorncom commented:
wat! ... ;-)  The token tests were all failing on my end when running `ember t -s`  Seems like the same issue came up on Travis too?

acorncom commented:
I looked at it for a bit and thought "this could take a while".  The job I need this for doesn't have lots of extra cushion, so I decided to punt on it for a bit.  So no, hadn't tracked down why things were breaking

wayne-o commented:
I think we are going to need some help with this as I don't understand addons enough and the failing tests are a bit vague for me to get to the bottom of :/

acorncom commented:
@jpadilla If you've got time to review this, would love to get your input on the errors we're getting here ...

RuslanZavacky commented:
@wayne-o actually you can, get this branch and test :) for me its working fine right now. Just spending more time validating everything else.

Commit:
#100 Added missing passwordField configuration option

mmahalwy commented:
is this good to go?

Commit:
#100 Refresh token expiration date in seconds left

acorncom commented:
@RuslanZavacky No, I don't have tests of the token refresh system yet, as our app is still in process of being launched and we're focusing on the frontend first :-)  Thanks for the heads up ...

acorncom commented:
**Background**
We store a few additional properties (in addition to `token`) on the result we return from the server to speed up lookups / setup.

Things I needed to change:
- changed `this.session.get(<property>)` to `this.get('session.data.authenticated.property')`.  Previously you had access to a raw JSON object directly on the session, but it now runs through a computed property which changes how we access things
- `config/environment.js#simple-auth` config key was changed to `ember-simple-auth`
- test helpers all need to be imported into each acceptance test using them and require the application to be passed in to each test helper

A few items we might want to change:
- old authenticators / authorizers were labeled `simple-auth-authenticator:jwt` or `simple-auth-authorizer:token`.  The new format is just to call it `authenticator:jwt` or `authorizer:token`
- might want to change our `config/environment.js` key to match by calling it `ember-simple-auth-token`
- might want to rename the npm package to match by calling it `ember-simple-auth-token`

Additional details on the upgrade can be found here (and most of the steps need to be followed): http://log.simplabs.com/post/131698328145/updating-to-ember-simple-auth-10

Commit:
#100 Unified naming ember-simple-auth-token

Authorizers, Authenticators do not have module prepended, now its simply:

authenticator:jwt
authenticator:token
authorizer:token

RuslanZavacky commented:
@acorncom unified naming across addon :) Renaming github package is up to @jpadilla. But yea, I agree that it should follow new standards, dropping -cli :)

RuslanZavacky commented:
I've found one bug with refresh token, if I am lucky, I'll fix it today, if not, then only on Monday :) stay tuned, but unless its fixed, this branch is't good to go :)

Commit:
#100 Fixed Refresh Ability & Tests

Commit:
#100 Switch to Phantomjs 2.0

Commit:
#100 Fixed async problems with tests

Commit:
#100 Removed time denomination from ms to s

Commit:
#100 timeFactor improvements

##100 Expiration back to its previous state

jpadilla commented:
@RuslanZavacky I'm taking some time tonight or in the weekend to take this out for a spin. Great work for putting all this together.

coladarci commented:
@RuslanZavacky - I just saw your comment from above - I am strongly against this change as it goes against the JWT spec. If I'm looking at the wrong spec (https://tools.ietf.org/html/rfc7519#section-4.1.4), or I'm missing something, please let me know, but this if this update ships, I think all server-side jwt solutions will have to be hand-rolled. A serious deal-breaker for most users, I'd imagine.

2468ben commented:
I also agree that however wrong/not wrong the spec is, the main bottleneck for folks right now is a tagged  version that's ember-simple-auth 1.0 compatible. So a PR with the breaking changes would let the compatible version come out, then another version with breaking changes could come when everyone agrees on those.

2468ben commented:
Also are the client & server not already communicating time as UTC? I thought that's how it worked to avoid the timezone issue.

coladarci commented on addon/authenticators/jwt.js:
```diff
@@ -168,7 +169,7 @@ export default TokenAuthenticator.extend({
           var token = response[_this.tokenPropertyName],
             tokenData = _this.getTokenData(token),
-            expiresAt = tokenData[_this.tokenExpireName],
+            expiresAt = _this.getCurrentTime() + tokenData[_this.tokenExpireName] * _this.timeFactor,
```
this is still wrong, right @RuslanZavacky ? As discussed, the token will contain the expiration time, not how long from now it expires.

coladarci commented on addon/authenticators/jwt.js:
```diff
@@ -240,10 +237,10 @@ export default TokenAuthenticator.extend({
-            tokenExpireData = {};
+          var token = response[_this.tokenPropertyName];
+          var tokenData = _this.getTokenData(token);
+          var expiresAt = _this.getCurrentTime() + tokenData[_this.tokenExpireName] * _this.timeFactor;
```
this is still wrong, right @RuslanZavacky ? As discussed, the token will contain the expiration time, not how long from now it expires.

RuslanZavacky replied:
I am still in progress. basically almost reverted whole this file.

Commit:
#100 Expiration back to its previous state

Commit:
#100 More flexible authentication data

cliedelt commented:
Any Update?

cliedelt commented:
@RuslanZavacky :+1:

mvrkljan commented:
@jpadilla any info on when we might expect this to be merged and available via cli?

jpadilla commented on addon/authenticators/jwt.js:
```diff
@@ -103,7 +104,7 @@ export default TokenAuthenticator.extend({
 
     return new Ember.RSVP.Promise(function(resolve, reject) {
-      var now = (new Date()).getTime();
+      var now = _this.getCurrentTime();
```
@RuslanZavacky Why is `getCurrentTime()` here useful? Why would you override it?

RuslanZavacky replied:
Abstraction for tests, we need to have defined state. Time is always random state in tests, which is not good.

jpadilla commented on addon/authenticators/token.js:
```diff
@@ -146,10 +146,16 @@ export default Base.extend({
-    authentication[this.passwordField] = credentials.password;
-    authentication[this.identificationField] = credentials.identification;
-    return authentication;
+    var password = credentials.password;
```
@RuslanZavacky Any specific reason for rewriting this block? Not a big fan of modifying the `credentials` object that's passed in by doing `delete ...`.

RuslanZavacky replied:
good question, have no idea what I was doing here. Maybe too much wine :)

jpadilla commented:
@RuslanZavacky left some inline comments. Also noticed something funky running the demo app.
1. Login
2. Wait for token refresh to start running
3. Go to another tab for x seconds
4. Endpoint returned a 500 error because token expired. This doesn't seem to happen if I stay on the same tab.

Thoughts on what might be going on? If I recall something like this happened at some point before. @erichonkanen do you recall what happened then?

Commit:
#100 getAuthenticateData to its previous state

acorncom commented:
Are we ready to merge this at this point?  I can help with further tidy up if needed in the next week or so ...

jpadilla commented:
@acorncom talked to @RuslanZavacky about checking out https://github.com/jpadilla/ember-cli-simple-auth-token/pull/100#issuecomment-160284249. That's all that's left that I can see.

RuslanZavacky commented:
I'll check it on weekend. Currently attending conference :)

Commit:
#100 Ability to pass additional headers to requests

Commit:
#100 Demo Application 1.0

Commit:
Merge remote-tracking branch 'origin/master' into simple-auth-update

; Conflicts:
;	tests/dummy/app/routes/index.js

RuslanZavacky commented:
@jpadilla I've fixed Demo app. Seems like it did not work before either?

Also I've added ability to pass additional headers for authentication request, useful, when you attach social account for already authenticated user :)

jpadilla commented:
@RuslanZavacky awesome, although I'm still seeing the issue with the expiring tokens. Managed to [record](https://monosnap.com/file/L40k1OgeV36t4JzYcrndfQut1vINzE) it.

RuslanZavacky commented:
@jpadilla hm, it might make sense. When you switch tab, Chrome stops execution, so refresh token can't be obtained. Token life is 10 seconds, refresh leeway is 5, so if you did not switch back in ~ 5 seconds, token expires. And if token is expired, you can't refresh it, you need to obtain the new one.

jpadilla commented:
@RuslanZavacky yeah that's what I thought. Could we handle the expiration gracefully on the dummy endpoint and have it return a valid response?

jrjohnson commented:
I think this has been an existing issue. We have a really big refresh leeway in our app to help with this.

RuslanZavacky commented:
@jpadilla I'll take a look :)

jpadilla commented:
@jrjohnson thanks for confirming

RuslanZavacky commented:
@jpadilla I think it should be a separate ticket, seems like there is no 'simple fix'

jpadilla committed:
Use expiresIn instead of expiresInSeconds

jpadilla committed:
Handle expired tokens

jpadilla committed:
Use Object.keys instead of Ember.keys

andruby commented on addon/authorizers/token.js:
```diff
@@ -73,27 +75,14 @@ export default Base.extend({
-      }
-
-      jqXHR.setRequestHeader(this.authorizationHeaderName, token);
+      block(this.authorizationHeaderName, this.authorizationPrefix + token);
```
Is it possible that this causes a regression when `authorizationPrefix` is set to `null`?

Using this branch and `authorizationPrefix: null` the authorization headers generated are `nullYekatoken..`.

RuslanZavacky replied:
yeap, you are right. Need to fix that :)

andruby replied:
Thank you for your quick & swift solution!

Commit:
#100 authorizationPrefix can be null

Commit:
Merge pull request #2 from acorncom/handle-expired-token

Improvements to dummy app

listepo commented:
Ping

jpadilla commented:
I'll be working on updating ember-cli and a few things before releasing a new version. Thanks everyone and specially @RuslanZavacky for all your awesome work.

RuslanZavacky commented:
:+1:

mmahalwy commented:
woot!
@coladarci

richardvenneman commented:
:pray:

jpadilla commented:
@RuslanZavacky @acorncom are there any documentation updates we need to do?

jpadilla commented:
Just published v1.0.0 https://www.npmjs.com/package/ember-simple-auth-token
